app/controllers/subjects_controller.rb

```diff
@@ -56,6 +55,40 @@ def update
 
   private
 
+  def can_update_subject
```

Comment:
This logic belongs in the model, first look at the comments below and then you can add this to Subject model

```ruby
def save_with_assignment_validation
  removed_skill_ids = @subject.assignments.filter(&:marked_for_destruction?).map(&:skill_id)
  if removed.skill_ids.any(|id| self.has_grades_in_skill? id)
    # Add errors here and return false
  else
    return save
  end
end
```